executor/aggregate.go

```diff
@@ -386,32 +389,134 @@ func (w *HashAggPartialWorker) shuffleIntermData(sc *stmtctx.StatementContext, f
+		err = item.VecEvalReal(ctx, input, buf)
+		if err != nil {
+			return err
+		}
+	case types.ETDuration:
+		err = item.VecEvalDuration(ctx, input, buf)
+		if err != nil {
+			return err
+		}
+	case types.ETDatetime, types.ETTimestamp:
+		err = item.VecEvalTime(ctx, input, buf)
+		if err != nil {
+			return err
+		}
+	case types.ETString:
+		err = item.VecEvalString(ctx, input, buf)
+		if err != nil {
+			return err
+		}
+	case types.ETJson:
+		err = item.VecEvalJSON(ctx, input, buf)
+		if err != nil {
+			return err
+		}
+	case types.ETDecimal:
+		err = item.VecEvalDecimal(ctx, input, buf)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		// This check is used to avoid error during the execution of `EncodeDecimal`.
-		if item.GetType().Tp == mysql.TypeNewDecimal {
-			v.SetLength(0)
+	}
+	// This check is used to avoid error during the execution of `EncodeDecimal`.
+	if item.GetType().Tp == mysql.TypeNewDecimal {
+		numRows := input.NumRows()
+		d64s := buf.Decimals()
+		for i := 0; i < numRows; i++ {
+			d64s[i].SetPrecision(0)
 		}
-		w.groupValDatums = append(w.groupValDatums, v)
 	}
-	var err error
-	w.groupKey, err = codec.EncodeValue(sc, w.groupKey[:0], w.groupValDatums...)
-	return w.groupKey, err
+	groupKey, err = codec.VectorizedEncodeValue(sc, groupKey, buf, tp)
+	return err
+}
+
+var (
+	defaultChunkSize  = 1024
+	partialResultPool = sync.Pool{
+		New: func() interface{} {
+			return make([][]aggfuncs.PartialResult, defaultChunkSize)
+		},
+	}
+	groupKeyPool = sync.Pool{
+		New: func() interface{} {
+			return make([][]byte, defaultChunkSize)
+		},
+	}
+)
+
+func allocPartialResultSlice(n int) [][]aggfuncs.PartialResult {
+	if n > defaultChunkSize {
+		return make([][]aggfuncs.PartialResult, n)
+	}
+	return partialResultPool.Get().([][]aggfuncs.PartialResult)
+}
+
+func deallocatePartialResultSlice(partialResult [][]aggfuncs.PartialResult) {
+	if cap(partialResult) <= defaultChunkSize {
+		partialResultPool.Put(partialResult)
+	}
+}
+
+func allocGroupKeySlice(n int) [][]byte {
+	if n > defaultChunkSize {
+		return make([][]byte, n)
+	}
+	return groupKeyPool.Get().([][]byte)
+}
+
+func deallocateGroupKeySlice(groupKey [][]byte) {
+	if cap(groupKey) <= defaultChunkSize {
+		groupKeyPool.Put(groupKey)
+	}
 }
 
-func (w baseHashAggWorker) getPartialResult(sc *stmtctx.StatementContext, groupKey []byte, mapper aggPartialResultMapper) []aggfuncs.PartialResult {
-	partialResults, ok := mapper[string(groupKey)]
-	if !ok {
-		partialResults = make([]aggfuncs.PartialResult, 0, len(w.aggFuncs))
-		for _, af := range w.aggFuncs {
-			partialResults = append(partialResults, af.AllocPartialResult())
+func (w baseHashAggWorker) getPartialResult(sc *stmtctx.StatementContext, groupKey [][]byte, mapper aggPartialResultMapper) [][]aggfuncs.PartialResult {
+	numRows := len(groupKey)
+	partialResults := allocPartialResultSlice(numRows)
+	partialResults = partialResults[:0]
+	defer deallocatePartialResultSlice(partialResults)
```

Comment:
We can't recycle `partialResults` here since it is still active then and it is returned and will be used by other functions.